src/mlpack/tests/lsh_test.cpp

```diff
@@ -917,4 +917,72 @@ BOOST_AUTO_TEST_CASE(MoveOperatorTest)
   CheckMatrices(distances, distances2);
 }
 
+/**
+ * Run LSH on (identical) dense and sparse data, making sure that we get the
+ * same results.  Note that the sparse data we are using isn't really
+ * sparse---the idea is to test that the class works correctly when the type is
+ * arma::sp_mat.
+ */
+BOOST_AUTO_TEST_CASE(SparseLSHTest)
+{
+  // kNN and LSH parameters (use LSH default parameters).
+  const int k = 5;
+  const int numTables = 5;
+  const int numProj = 2;
+  const double hashWidth = 50.0;
+  const int secondHashSize = 99901;
+  const int bucketSize = 500;
+
+  // Read iris training and testing data as reference and query sets.
+  const string trainSet = "iris_train.csv";
+  const string testSet = "iris_test.csv";
+  arma::mat rdata;
+  arma::mat qdata;
+  data::Load(trainSet, rdata, true);
+  data::Load(testSet, qdata, true);
+
+  // Run on dense data.
+  LSHSearch<> denseLSH(
+      rdata,
+      numProj,
+      numTables,
+      hashWidth,
+      secondHashSize,
+      bucketSize);
+
+  arma::Mat<size_t> denseNeighbors;
+  arma::mat denseDistances;
+  denseLSH.Search(qdata, k, denseNeighbors, denseDistances);
+
+  // Now create and run on sparse data.
+  arma::sp_mat sparseRData(rdata);
+  arma::sp_mat sparseQData(qdata);
+
+  LSHSearch<NearestNeighborSort, arma::sp_mat> sparseLSH(
+      sparseRData,
+      denseLSH.Projections(),
+      hashWidth,
+      secondHashSize,
+      bucketSize);
+
+  arma::Mat<size_t> sparseNeighbors;
+  arma::mat sparseDistances;
+  sparseLSH.Search(sparseQData, k, sparseNeighbors, sparseDistances);
+
+  BOOST_REQUIRE_EQUAL(denseNeighbors.n_rows, sparseNeighbors.n_rows);
+  BOOST_REQUIRE_EQUAL(denseNeighbors.n_cols, sparseNeighbors.n_cols);
+  BOOST_REQUIRE_EQUAL(denseDistances.n_rows, sparseDistances.n_rows);
+  BOOST_REQUIRE_EQUAL(denseDistances.n_cols, sparseDistances.n_cols);
```

Comment:
We only compare the dimensions but not the actual values?

Reply:
Yeah---the reason for this is that LSH is randomized and points may fall into different buckets, so the returned neighbors may be completely different each time.  Since we don't really have any performance guarantees, I figured it was sufficient to simply check that results are returned at all.  Realistically if all the Armadillo sparse computations work the same way as the dense ones (which they do; that's tested in the Armadillo codebase), then this should be exactly the same.  I would say the primary reason for the test is just to make sure that `LSHSearch` with `sp_mat` even compiles. :)

Reply:
Thanks for the clarification, missed that LSH is randomized, makes sense.